tools/autograd/derivatives.yaml

```diff
@@ -1129,9 +1129,9 @@
   grad_output: softplus_backward(grad, self, beta, threshold, output)
   self: softplus_double_backward(grad * grad_output, self, beta, threshold)
 
-- name: softmax_backward_data(Tensor grad_output, Tensor output, int64_t dim, Tensor self)
-  grad_output: softmax_backward_data(grad, output, dim, self)
-  self: softmax_double_backward(grad, grad_output, dim, output)
+- name: _softmax_backward_data(Tensor grad_output, Tensor output, int64_t dim, Tensor self)
+  grad_output: _softmax_backward_data(grad, output, dim, self)
+  self: softmax_double_backward(grad, grad_output, dim, output).type_as(self)
```

Comment:
Hmm so it's really mostly targeted for inference and we're falling back to a slower pathway in the backward? 😕

Reply:
No, these are derivatives for double backward, so just for half training w/o double backward both forward and backward will use faster kernels, and avoid cast before softmax/after softmax backward. Double backward for mixed precision will be on fp32 tensors, true, but that's the same as converting input to fp32 and using fp32 softmax all around, so it's never worse than what we have currently.